#275 update to HDI and holochain 0.0.152

Connoropolous opened the pull request:
In case not using nix-shell, these are the package versions for this branch:
cargo install holochain --version 0.0.152
cargo install holochain_cli --version 0.0.50
cargo install lair_keystore --version 0.0.11

additionally you will need the following:
- Node v16
- pnpm
- install the `wasm32-unknown-unknown` target: `rustup target add wasm32-unknown-unknown`

Current HDK: 0.0.144

closes #275 

- [x] refactor bundles/dna_templates to have 'integrity' and 'coordinator' zomes
- [x] add updated dna-auth-resolver
- [x] set up nix-shell for holochain v0.0.150
- [x] update HDK versions to 0.0.142, instead of anywhere that it has 0.0.136 set, fix compilation issues
- [x] rename any reference from `Header` to `Action`
- [x] rename any reference from `Element` to `Record`
- [x] introduce entry enum type
- [x] separate out an integrity zome for each zomes
- [x] work those integrity zomes into the bundles/dna_templates
- [x] update @holochain/client (ask Connor about this)
- [x] define `IndexSegment` entry type and its method implementations in the integrity zome, or remove
- [x] define UUID entry type and wrapper enum in integrity, or remove (working with hashes only)
- [x] define separate link types in `hdk_records`, along with integrity zome crate (or define in link type in each object type integrity zome and pass in the type to hdk_records functions)
- [x] fix the integration tests

Commit:
revising dna bundles for integrity zomes and new dna-auth-resolver dependency

Commit:
Merge branch 'sprout' into feature/275-update-holochain-150

Connoropolous commented:
@pospi  FYI @weswalla is back and they're taking over this from me

Connoropolous commented:
we've just landed the migration on the Acorn project, so have learned a few lessons along the way and feel prepared for this

weswalla committed:
update nix config to holochain 150

weswalla committed:
update hdk_records crate hdk dependency to 142

weswalla committed:
rename element to record in hdk_records crate

weswalla committed:
rename header to action in hdk_records crate

weswalla committed:
create and update entry input type updated in hdk_records crate

weswalla committed:
define link type and use when creating and getting links in hdk_records crate

weswalla committed:
use TypedPath in hdk_records crate

weswalla committed:
point to proper holonix revision in default.nix

weswalla committed:
separate out integrity types into separate crate

weswalla committed:
header with action in hdk_semantic_indexes crate

weswalla committed:
add integrity_types crate to cargo workspace manifest

weswalla committed:
disambiguate holo_hash import

weswalla committed:
change link type name from HdkRecordsLinkTypes to LinkTypes

weswalla committed:
use link type in paths in hdk_semantic_indexes

weswalla committed:
update hdk_time_indexing to hdk 142

weswalla committed:
disambiguate holo_hash import in hdk_uuid_types

weswalla committed:
revert hdk and holo_hash crate imports to ambiguous version

weswalla committed:
revert holo_hash import in hdk_uuid_types to original format

weswalla committed:
use hdi instead of hdk in integrity_types crate

Commit:
update holochain-client to fork for now

weswalla committed:
Merge branch 'feature/275-update-holochain-150' of github.com:h-REA/hREA into feature/275-update-holochain-150

weswalla committed:
generate EntryTypes enum and wrap entry struct in enum when calling create_entry

weswalla committed:
bumped all hdk dependencies to 0.0.142

weswalla committed:
deleted unused integrity_types module from hdk_record

weswalla committed:
added agent integrity zome

weswalla commented on zomes/rea_agent/integrity_zome/src/lib.rs:
```diff
@@ -0,0 +1,12 @@
+ * @package Holo-REA
+ */
+
+pub use hc_zome_rea_agent_storage::EntryTypes;
+pub use integrity_types::LinkTypes;
```
does this work? or do the procedural macros (`#[hdk_entry_defs]` etc) need to be invoked in this crate? @pospi @Connoropolous

weswalla replied:
I suppose a different approach would be to define the `EntryTypes` enum in this file, and implement the `From<EntryStorage>` trait directly in here. I think this makes more sense actually.

Connoropolous replied:
FYI @pospi 
Wesley and I jammed on this today and came up with some strategies

Connoropolous commented:
@weswalla on this branch (just mirroring a change I made on `sprout` branch) delete the whole `_templates` and `example` folders? They're causing conflicts now, because I deleted them, but you changed them. Just delete them and we're fine.

weswalla committed:
removed deprecated field from EntryDef

weswalla committed:
use LinkTypes defined in integrity zome in get_links calls

weswalla commented:
Does this general integrity zome pattern look right?

weswalla committed:
removed _templates and example directories

Commit:
separate application zomes from utility zomes in integrity zome definition

Commit:
added agreement integrity zome crate

Connoropolous commented on bundles/dna_templates/agent/dna.yaml:
```diff
@@ -26,11 +26,13 @@ integrity:
-      bundled: "<repository-path>/target/wasm32-unknown-unknown/release/hc_zome_rea_agent_integrity.wasm"
 
 coordinator:
   zomes:
```
You're going to need to add something like this, similar to how it is done for `remote_auth`
```
      dependencies:
        - name: agent_integrity
```

Connoropolous commented on zomes/rea_agreement/integrity_zome/src/lib.rs:
```diff
@@ -0,0 +1,12 @@
+ * @package Holo-REA
+ */
+
+pub use hc_zome_rea_agent_storage::EntryTypes;
```
```suggestion
pub use hc_zome_rea_agreement_storage::EntryTypes;
```

Commit:
Merge branch 'sprout' into feature/275-update-holochain-150

Connoropolous commented on bundles/dna_templates/agent/dna.yaml:
```diff
@@ -1,34 +1,50 @@
+    # application zomes
+    - name: agent
+      bundled: "<repository-path>/target/wasm32-unknown-unknown/release/hc_zome_rea_agent.wasm"
+    - name: agent_index
```
For all of this to work, there might need to be an 'index_integrity' zome as well, that gets added as a dependency to all the index zomes. It might be able to be the same zome for all of them

weswalla replied:
are there any new integrity types that would need to get defined though? Can we not just import the `LinkTypes` from the same integrity zome as the `agent` zome does?

weswalla replied:
I guess I was thinking we only need to define 1 integrity zome per dna, but if we are to split every zome into coordinator and integrity then we'd have 2

Connoropolous replied:
Ah, I see. that could be true. Depends if the index zomes need entry types, or just link types. And if we were to make LinkTypes more specific then we'd need more specific integrity zomes. 

Your proposal could work

Commit:
define coordinator zome dependencies in dna.yaml

weswalla committed:
fix crate name import

weswalla committed:
Merge branch 'feature/275-update-holochain-150' of github.com:h-REA/hREA into feature/275-update-holochain-150

weswalla committed:
use new wasm_error! macro

weswalla committed:
extended trait bounds to include entry type wrapper enum, and handler error type conversion

weswalla committed:
moved index_integrity_zome to zome directory and added link type variants

weswalla committed:
use new integrity zome pattern in agent dna

weswalla committed:
use integrity link types in hdk_semantic_indexes

weswalla committed:
use integrity zome for agreement dna

weswalla committed:
use integrity zomes for observation dna

weswalla committed:
use integrity zomes for plan dna

weswalla committed:
move semantic index integrity crate to hdk_semantic_indexes dir

weswalla committed:
define EntryTypes enum inside integrity zome along with conversion trait

weswalla committed:
use integrity zomes for planning dna

weswalla committed:
use integrity zomes for proposal dna

weswalla committed:
use integrity zomes for specification dna

weswalla committed:
add all new crates to root cargo.toml

weswalla committed:
replaced Header with Action

weswalla committed:
replaced Element with Record

weswalla committed:
removed required_validation_type field from EntryDef structs

weswalla committed:
removed index segment entry definition from semantic indexes integrity zome

weswalla committed:
use updated version of dna_auth_resolver

pospi committed:
Merge branch 'sprout' into feature/275-update-holochain-150

pospi committed:
update all HDK refs to latest version

pospi committed:
update Cargo lockfile

pospi committed:
update PNPM lockfile

pospi committed:
update hdi & holo_hash dependencies to latest

pospi committed:
update lockfile

pospi committed:
update thirdparty git dependencies

pospi committed:
fix renames broken in merge f1db2bed

pospi committed:
more SignedHeaderHashed -> SignedActionHashed renames

pospi committed:
update Entry creation & update logic to handle integrity zome type conversions

EntryDefId is now specified by type signature & conversion traits

entry_def_id parameter no longer needed in entry writes- removed from entry methods & record update, but still needed in record creation for index creation until index zomes refactor is completed

pospi committed:
update all calls to update_record to remove old EntryDefId parameter

pospi committed:
update anchored record helpers to pass link types through API

hardcode link types for semantic indexes lib via its dedicated integrity zome

removed link_helpers from hdk_records and inlined relevant code into anchored_record_helpers & semantic indexes lib

update for new Path usage to not store to DHT

fix link type imports location for semantic indexes zome macro

pospi committed:
update lockfile

pospi committed:
update Unit zome logic to use declared UnitIdentifier integrity link type

pospi committed:
remove unneeded record type ID from identifier handling methods, inference is now handled by LinkType

pospi committed:
update Unit update & read methods to remove unneeded EntryDefId param

pospi committed:
update all zome calling logic to omit EntryDefId param for reads & updates

pospi committed:
fix missing imports in integrity zomes & ambiguous name in Unit integrity zome

pospi committed:
fix error handling in Action zome

pospi committed:
fix caller error about ambiguous trait bounds which would mandate defining all types

pospi committed:
change Addressable types to be hashable as entry, but not storable; and remove now unneeded storage of index pointers

pospi committed:
fix duplicate derivation of trait on record entries

pospi committed:
remove ensure() method for IndexSegment so not stored to DHT

pospi committed:
remove traits needed for storage of Addressable pointers on DHT, now references EntryHash directly as much simpler and HDI means we can link to remote data

also fixes some trait ambiguity in entry creation due to inability to disambiguate WasmError types

pospi committed:
remove unneeded entry conversion traits and entry type IDs from indexing lib

pospi committed:
add index integrity zome dependency to all indexing zomes

pospi committed:
remove old EntryDefs from all index zomes, now serviced by dependency in integrity crate

pospi committed:
update lockfile

pospi committed:
remove more Addressable entry conversion traits & params from indexing lib

pospi committed:
add parameterisation to all record creation methods to specify EntryType to use

pospi committed:
fix unneeded use of typed paths in String-based indexes causing error

pospi committed:
remove more unneeded trait bounds

pospi committed:
remove entrydefs from coordinator zomes, update record read API calls, update validations for HDI

pospi committed:
parameterise LinkTypes for Unit record reads & updates

pospi commented:
Made a lot of progress, but a couple of new problems left to resolve:

- [x] `hc_zome_rea_unit_lib` isn't compiling and complains that its `LinkType`s aren't convertible to `ScopedZomeType<LinkType>`. I think this might be due to incorrect trait bounds in `anchored_record_helpers.rs`.
- [x] coordinator zomes which depend both on their own integrity zome crates *and* indirectly on `hc_zome_dna_auth_resolver_integrity` are causing linker errors about duplicate definitions of `entry_defs`. Two possible solutions, both requiring changes to the lib which you are welcome to make on its main branch:
    - create some intermediary crate defining the storage structs and declare them in each type-specific integrity zome in which they're needed (eg. `hc_zome_rea_resource_specification_integrity::EntryTypes`). In this case you will probably need to parameterise `hdk_records::rpc::call_zome_method` and its dependencies/dependants in order to specify the entry type struct.
    - **or** just make changes to `dna-auth-resolver` to use the built-in capability storage functionality of the HDK and do away with its integrity zome entirely.

Commit:
Merge branch 'sprout' into feature/275-update-holochain-150

pospi committed:
update hdk_records identity handling to presume & infer local IDs when reading, obviating need for DHT ops

pospi committed:
rewrite identity handling for remote entries in semantic indexes to store in a self-referential link tag

Commit:
Merge branch 'sprout' into feature/275-update-holochain-150

pospi committed:
write identifier data to link tags only if not already written

pospi committed:
ensure fully-qualified ID tags are written to all indexed entrry addresses

fully-qualified local zome links are probably not necessary, but less refactoring to leave as-is for now

partial revert of 28013478

pospi committed:
Merge branch 'feature/275-update-holochain-150' of github.com:h-REA/hREA into feature/275-update-holochain-150

Commit:
get unit compiling

weswalla committed:
use updated version of dna_auth_resolver with integrity/core split

weswalla committed:
removed entry def callback from unit coordinator zome

weswalla committed:
WIP: pass in entry type and link type flags to avoid multiple integrity dependency errors

weswalla commented:
the most recent commit (https://github.com/h-REA/hREA/pull/340/commits/e873e797ea545f9301b0226374ab46f9d4a76341) makes all crate compile, however, I think we will want to replicate the core/integrity split pattern of dna_auth_resolver (https://github.com/holochain-open-dev/dna-auth-resolver/blob/2c9951ae5cc567b7560773d94ec6c34a0afe0758/integrity/src/lib.rs#L9) because right now the `__num_entry_types` and `__num_link_types` callbacks are left out of the 2 crates listed below which I'm assuming will be an issue on the holochain side of things.

There are two places we will need to handle this:
1. `hc_zome_rea_agent_integrity` crate
2. `hc_zome_rea_economic_event_integrity` crate

With (1), the error occurs because we define a link type in the agent integrity zome, but the `hdk_semantic_indexes_integrity` crate already defines a `LinkTypes`. With (2), the error occurs because the `hc_zome_rea_economic_event_lib` crate depends on the `hc_zome_rea_economic_resource_lib` crate which leads to two separate entry type definitions using the `#[hdk_entry_defs]` macro.

One solution for (1) is to split the `hdk_semantic_indexes_integrity` crate into 2 crate, like in https://github.com/holochain-open-dev/dna-auth-resolver/blob/2c9951ae5cc567b7560773d94ec6c34a0afe0758/integrity/src/lib.rs#L9, however, this wouldn't address (2).

The other solution (that I can think of) is to do the split on the `zome` side rather than the `lib` side. Meaning both `hc_zome_rea_agent_integrity` and `hc_zome_rea_economic_event_integrity` are decomposed into 2 crates each (core/integrity), thus solving (1) and (2).

Commit:
replaced uid with network_seed and fixed other small dna config issue
separated entry type structs and link structs from the callbacks
integrity zomes shouldn't be used as imported crates

Connoropolous commented:
I did the split on both sides, semantic and classes and for the classes I did all of them instead of being selective

Connoropolous commented:
Next step is integration tests I think. Seeing how much we broke along the way

Connoropolous commented on lib/hdk_indexing_core/src/lib.rs:
```diff
@@ -0,0 +1,20 @@
+//     hdk_semantic_indexes_core::LinkTypes::len()
+// }
+
+#[hdk_link_types(skip_no_mangle = true)]
+pub enum LinkTypes {
+    IdentityAnchor,
+    SemanticIndex,
+    TimeIndex,
+}
```
@pospi I'm moving your comment to the PR, as it's easier to track:

Pospi said
"hmm, I don't think there should be any situations where all 3 of these link types should be managed by the same controller zome. If this is just for Agent then it's possibly an architectural problem to resolve. I've been working steadily to separate those crates, ideally soon we can drop the semantic indexes dependency on hdk_records entirely and that will reduce this kind of confusion."

Connoropolous replied:
My response...

But semantic and time indexing do get used in a colocated context right? So is it IdentityAnchor that would get isolated out?

pospi replied:
yep! thats right

Connoropolous replied:
I'm going to leave it as a TODO I think

pospi replied:
addressed this now, it's gone from the lib

Connoropolous commented:
Ok so, 

```
cd test
TRYORAMA_LOG_LEVEL=debug WASM_LOG=debug RUST_LOG="debug,wasmer_compiler_cranelift=error,holochain::core::workflow=error" RUST_BACKTRACE=1 npx tape agent/**/*.js
```

results in all failures, and a wasm related 'host/guest' interface issue:

```
error during scenario.addPlayerWithHappBundle:  {
  type: 'error',
  data: {
    type: 'internal_error',
    data: 'Conductor returned an error while using a ConductorApi: RibosomeError(WasmRuntimeError(RuntimeError { source: User(WasmError { file: "/Users/connor/.cargo/registry/src/github.com-1ecc6299db9ec823/holochain-0.0.152/src/core/ribosome/real_ribosome.rs", line: 416, error: Compile("Error while importing \\"__wbindgen_placeholder__\\".\\"__wbindgen_describe\\": unknown import. Expected Function(FunctionType { params: [I32], results: [] })") }), wasm_trace: [], native_trace:    0: <unknown>\n'
```

looking into the line throwing it:
https://github.com/holochain/holochain/blob/develop/crates/holochain/src/core/ribosome/real_ribosome.rs#L414

Connoropolous commented:
Oh! Damien posted about this issue recently and discovered the fix. Woohoo!
https://discord.com/channels/727959104777224303/728647693244694648/1005632262865432586
https://discord.com/channels/727959104777224303/728647693244694648/1005668887431487488

Commit:
use Damiens bug fix for chrono 0.4.7 having introduced wasm_bindgen as a dependency

Connoropolous commented:
Pushed the fix, that is letting us get the integration tests running

Commit:
update tryorama dependency

Commit:
add breaks to the integration tests that display the filename and separate the test running dots, for context

weswalla committed:
use IdentityAnchor link type variant in string_index_hash to fix agent integration test errors

weswalla committed:
change indexing order for accessing contained resources to make tests pass

weswalla commented:
It seems like the remaining errors (there may be others I haven't investigated yet) are related to `dna_auth_resolver`, as the following error is consistent across tests:
```
Err(Wasm(WasmError { file: "/Users/weswalla/.cargo/git/checkouts/dna-auth-resolver-bb2f8d37890db247/2c9951a/core/src/lib.rs", line: 32, error: Guest("AvailableCapability does not map to any ZomeId and LinkType that is in scope for this zome.") }))])
```

which is related to https://github.com/holochain-open-dev/dna-auth-resolver/blob/2c9951ae5cc567b7560773d94ec6c34a0afe0758/core/src/lib.rs#L32

where the error in generated in holochain: https://github.com/holochain/holochain/blob/29f7cf536c030c5d79c6595f4bdbcb0c61c3b9d1/crates/hdk_derive/src/link_types.rs#L71

Commit:
use generics for call_zome

Commit:
replace DataIntegrityError with WasmError in the conversion traits

Commit:
generic type annotations and other manage_index fixes

Commit:
Merge branch 'sprout' into feature/275-update-holochain-150

Commit:
add the missing __num_link_types externs

Commit:
Update default.nix

Commit:
re-enable validation by moving to integrity zomes

Commit:
fix the bad integrity zomes for indexing assumption (read more in desc)

we had a bad assumption in play, which is that the relationship between the indexing integrity zomes could be one-to-many (one integrity zome, many indexing zomes), but that was bad. it meant that when calling get_links, links from other indexing zomes could be mixed in with the links for the intended indexing zome

Commit:
revert flipped values

Commit:
missed __num_link_types for economic resource

weswalla committed:
use real resource specification id when creating economic events

Commit:
change back to @holochain/client now they've fixed build issues

Commit:
cleanup console logs

Commit:
up 80 minutes to 100 minutes to let macos CI pass

pospi committed:
fix hardcoding of link type in internal method

pospi committed:
remove use of TypedPath in semantic index client lib, not necessary to commit it so can just use untyped Path to get the hash

pospi committed:
use different LinkType name for identities in semantic index lib to avoid ambiguity

Connoropolous commented:
merrrrrrrging
